### ISSUE TITLE
Add a pardisomkllib option

### ISSUE THREAD

**Issue by amontoison** [coin-or/Ipopt#654]
Is it possible to add an `pardisomkllib` like `hsllib` or `pardisolib`?
It's also not possible to switch to `pardisomkl` solver dynamically.
```shell
Setting: "pardisomkl" is not a valid setting for Option: linear_solver. Check the option documentation.

### linear_solver (String)  ###
Category: Linear Solver
Description: Linear solver used for step computations.
Valid Settings:
	ma27 (load the Harwell routine MA27 from library at runtime)
	ma57 (load the Harwell routine MA57 from library at runtime)
	ma77 (load the Harwell routine HSL_MA77 from library at runtime)
	ma86 (load the Harwell routine MA86 from library at runtime)
	ma97 (load the Harwell routine MA97 from library at runtime)
	pardiso (load the Pardiso package from pardiso-project.org from user-provided library at runtime)
	mumps (use the Mumps package)
	custom (use custom linear solver (expert use))
```
It should be great to also have a similar option for the Spral solver.

**Comment by svigerske** [comment]
Probably it is possible, but it is extra effort to implement this.

The reason to do so for HSL and Pardiso (from pardiso-project.org) was that these are codes that one cannot freely redistribute. But MKL and Spral have liberal licenses. So like you have been doing for Mumps, you can just link against these while building Ipopt.

With Pardiso from MKL, I'm also not sure how smooth things will work if one uses a different Blas/Lapack for Ipopt. Loading MKL libs with `RTLD_LOCAL` seems to be a problem due to the MKL distributing their symbols over different libraries. Loading with `RTLD_GLOBAL` would load also all the Blas/Lapack symbols into the main namespace, possibly creating new issues.

**Comment by amontoison** [comment]
If we build Ipopt with Mumps and Spral. Will we be able to switch between Mumps and Spral solvers?
If it's the case, it solves the problem for Spral.

For PardisoMKL, I don't think that we can easily solve the problem because we already compile Ipopt with LBT ([libblastrampoline](https://github.com/JuliaLinearAlgebra/libblastrampoline)) and we don't want to use the symbols of `libmkl_rt` for BLAS / LAPACK routines.
Maybe the order of the flags [here](https://github.com/JuliaPackaging/Yggdrasil/blob/master/C/Coin-OR/Ipopt/build_tarballs.jl#L28-L46) is relevant but I never tested.
I thought that loading PardisoMKL externally with `pardisomkllib` will avoid this issue.

**Comment by svigerske** [comment]
Yes, if you build Ipopt with Spral, the default for the linear solver in Ipopt will change to `spral`, but you can use Ipopt option `linear_solver` to select Mumps.

I would just leave this issue open for now as a reminder to add an option to also load MKL's Pardiso at runtime some day.

Btw, the line 45 (https://github.com/JuliaPackaging/Yggdrasil/blob/master/C/Coin-OR/Ipopt/build_tarballs.jl#L45) seems to be missing a `$` before `{BLAS_LAPACK}`.

**Comment by amontoison** [comment]
Thanks for the typo with `{BLAS_LAPACK}`!
I plan to recompile Ipopt tomorrow, do you have a flag / option to compile Ipopt with `-03`?
I didn't find it in the documentation where we should set the optimization level.

Off-topic: For information, we just released [JuliaHSL](https://licences.stfc.ac.uk/product/julia-hsl). :tada: 
It could be also relevant for users that are using Ipopt without Julia.

**Comment by svigerske** [comment]
You would have to overwrite or augment the compiler flags that configure is determining.
By default, that is `-02 -DNDEBUG` for C/C++ and `-O2` for Fortran.

You should be fine by adding `CFLAGS="-O3 -DNDEBUG" CXXFLAGS="-O3 -DNDEBUG" FFLAGS=-O3 FCFLAGS=-O3` to the configure flags.

There are also `ADD_XYZFLAGS` variables, which you could set to `-O3` and then rely on the compiler picking the latter `-O` compiler flag.

**Comment by amontoison** [comment]
> Yes, if you build Ipopt with Spral, the default for the linear solver in Ipopt will change to `spral`, but you can use Ipopt option `linear_solver` to select Mumps.
> 
> I would just leave this issue open for now as a reminder to add an option to also load MKL's Pardiso at runtime some day.
> 
> Btw, the line 45 (https://github.com/JuliaPackaging/Yggdrasil/blob/master/C/Coin-OR/Ipopt/build_tarballs.jl#L45) seems to be missing a `$` before `{BLAS_LAPACK}`.

@svigerske 
I recompiled Ipopt with SPRAL but the default solver is still MUMPS and we can't switch to the SPRAL linear solver.
It's fine for us to keep MUMPS as the default solver but we would like to switch to the SPRAL solver.
https://github.com/JuliaPackaging/Yggdrasil/pull/6836/files

**Comment by svigerske** [comment]
What you mean you "can't switch to the SPRAL linear solver"? Setting `linear_solver spral` should do it.

The default is currently still that MUMPS is prefered over SPRAL. I don't really know how good or stable SPRAL it. They didn't have a release for a long time that would have worked with Ipopt, until you pinged them. That they then just tagged immediately whatever there was and don't provide release notes doesn't really build up confidence on their release+test cycle. Some issue I reported is also still unanswered.

If you want to change the default in your link, then you can add a `ipopt->Options()->SetStringValue("linear_solver", "spral", true, true)` before the users options are set. You can use function [IpoptGetAvailableLinearSolvers()](https://coin-or.github.io/Ipopt/IpLinearSolvers_8h.html) to check whether Spral has been linked in, if that is necessary.

**Comment by amontoison** [comment]
> What you mean you "can't switch to the SPRAL linear solver"? Setting `linear_solver spral` should do it.

I have the following error:
```julia
# Use the linear solver SPRAL
julia> set_attribute(model, "linear_solver", "spral")

julia> optimize!(model)
Setting: "spral" is not a valid setting for Option: linear_solver. Check the option documentation.

### linear_solver (String)  ###
Category: Linear Solver
Description: Linear solver used for step computations.
Valid Settings:
	ma27 (load the Harwell routine MA27 from library at runtime)
	ma57 (load the Harwell routine MA57 from library at runtime)
	ma77 (load the Harwell routine HSL_MA77 from library at runtime)
	ma86 (load the Harwell routine MA86 from library at runtime)
	ma97 (load the Harwell routine MA97 from library at runtime)
	pardiso (load the Pardiso package from pardiso-project.org from user-provided library at runtime)
	mumps (use the Mumps package)
	custom (use custom linear solver (expert use))
Default: "mumps"
ERROR: IPOPT: Couldn't set option 'linear_solver' to value 'spral'.
```

**Comment by svigerske** [comment]
It seems that Ipopt believes that Spral hasn't been linked in. Check the configure output, then check that `IPOPT_HAS_SPRAL` is defined in `src/Common/config.h` and that you indeed use the libraries that you have build.

You could add something like
```
src/Apps/AmplSolver/ipopt --print-options | grep -A 20 linear_solver
```
to the build script to show the possible values for the linear_solver option as accepted by the Ipopt (AMPL-)executable.

**Comment by amontoison** [comment]
Thanks for the help Stefan, an old shared library `libipopt.so` was used last time....
I just opened a PR #677 to don't use spral as the default linear solver if we have more "stable" linear solvers available.

**Comment by amontoison** [comment]
@svigerske I worked on a shim library `libmkl_pardiso.$dlext` today that only contains the symbols for the linear solver "MKL Pardiso".
https://github.com/amontoison/MKL_Pardiso
It ensures that we can use a different library for BLAS/LAPACK symbols.  
This is what we want in Julia, where we rely on `libblastrampoline`.  
My question is: how do I provide `libmkl_pardiso` to Ipopt?  
It seems that we don't have a dedicated option for this linear solver.
Should I add `-lmkl_pardiso` with `-lblastrampoline` for the BLAS or LAPACK option?

**Comment by svigerske** [comment]
Yes, currently it is assumed that the MKL-Pardiso comes with Lapack, so configure checks for symbol `pardiso` in the Lapack library. If you add `-lmkl_pardiso` to your Lapack flags, that could work.

**Comment by rmok-git** [comment]
I have a successful build using intel mkl 2025.1 with ipopt 3.14.17 on ubuntu using gcc/gfortran with coinbrew:

`coinbrew build Ipopt --with-lapack-lflags='-L..../intel/oneapi/mkl/2025.1/lib -lmkl_rt -lgomp -Wl,-rpath -Wl,..../intel/oneapi/mkl/2025.1/lib' --without-hsl`

as long as I have mumps as the other solver. I tested with `ipopt Ipopt/test/mytoy.nl  linear_solver=pardisomkl`. I didn't test other mkl runtime setup.

I want to report that **_the build fails when using pardisomkl as the only solver_** (by adding --without-mumps to `coinbrew`).  Basically, test programs fail with

`..../Ipopt/3.14.17/no-mumps/Ipopt/3.14.17/test/.libs/hs071_cpp: symbol lookup error: .../intel/oneapi/mkl/2025.1/lib/libmkl_intel_thread.so.2: undefined symbol: omp_in_parallel`

Putting all the details here would be too much. Here is what I see in libipopt.la
```
# Libraries that this one depends upon.
dependency_libs=' -L..../intel/oneapi/mkl/2025.1/lib -lmkl_rt -ldl'
```

When building with mumps, `-lgomp` is captured in libcoinmumps.la, and only ThirdParty/Mumps does not label --with-lapack-lflags as unrecognized option during configuration.

**Comment by svigerske** [comment]
@rmok-git  I don't think this is related to this issue about adding a `pardisomkllib` option to Ipopt.

And I can't reproduce how the `-lgomp` got lost from your `--with-lapack-lflags`.
If I configure with `--with-lapack-lflags='-L.../intel/oneapi/mkl/2025.1/lib -lmkl_rt -lgomp -Wl,-rpath -Wl,..../intel/oneapi/mkl/2025.1/lib', I get
```
dependency_libs=' -L.../intel/oneapi/mkl/2025.1/lib -lmkl_rt -lgomp -ldl'
```
You're sure you didn't forget the `-lgomp` when disabling Mumps?

**Comment by rmok-git** [comment]
I ran my command with `/usr/bin/time -v` and capture everything in a log file. You can see the command I used, including the typo disable-slient-rules. `--without-mumps and --without-hsl` is at the end.

```
    ******** Test FAILED! ********
Test program existed with nonzero status.
Testing GetCurr Example...
0 
 ---- 8< ---- Start of test program output ---- 8< ----
/home/russell/nfs-builds/Ipopt/3.14.17/no-mumps/Ipopt/3.14.17/test/.libs/getcurr: symbol lookup error: /nfs/optimizer/tools/intel/oneapi/mkl/2025.1/lib/libmkl_intel_thread.so.2: undefined symbol: omp_in_parallel
 ---- 8< ----  End of test program output  ---- 8< ----
 
    ******** Test FAILED! ********
Test program existed with nonzero status.
make[1]: *** [Makefile:817: test] Error 1
make[1]: Leaving directory '/home/russell/nfs-builds/Ipopt/3.14.17/no-mumps/Ipopt/3.14.17/test'
make: *** [Makefile:852: test] Error 2
Command exited with non-zero status 2
	Command being timed: "/nfs/optimizer/tools/coin-or/coinbrew build Ipopt -j 4 --verbosity=4 --prefix=/home/russell/Ipopt/3.14.17/no-mumps --build-dir /home/russell/nfs-builds/Ipopt/3.14.17/no-mumps --disable-slient-rules ADD_CFLAGS=-pthread -fopenmp ADD_CXXFLAGS=-pthread -fopenmp -std=c++14 ADD_FCFLAGS=-fopenmp --with-lapack-lflags=-L/nfs/optimizer/tools/intel/oneapi/mkl/2025.1/lib -lmkl_rt -lgomp -Wl,-rpath -Wl,/nfs/optimizer/tools/intel/oneapi/mkl/2025.1/lib --with-lapack-cflags=-I/nfs/optimizer/tools/intel/oneapi/mkl/2025.1/include --without-mumps --without-hsl"
	User time (seconds): 66.16
	System time (seconds): 9.33
```

**Comment by amontoison** [comment]
`-lgomp` is probably not the good flag here because the Pardiso library was compiled with Intel OpenMP.
You need a `-liomp5` or something similar.

**Comment by rmok-git** [comment]
The intel installation changed because before I only install MKL without the compiler runtime. Still, `-liomp5` fails like before.

```
    ******** Test FAILED! ********
Test program existed with nonzero status.
Testing GetCurr Example...
0 
 ---- 8< ---- Start of test program output ---- 8< ----
/home/russell/nfs-builds/Ipopt/3.14.17/no-mumps/Ipopt/3.14.17/test/.libs/getcurr: symbol lookup error: /nfs/optimizer/intel/oneapi/mkl/2025.1/lib/libmkl_intel_thread.so.2: undefined symbol: omp_in_parallel
 ---- 8< ----  End of test program output  ---- 8< ----
 
    ******** Test FAILED! ********
Test program existed with nonzero status.
make[1]: *** [Makefile:817: test] Error 1
make[1]: Leaving directory '/home/russell/nfs-builds/Ipopt/3.14.17/no-mumps/Ipopt/3.14.17/test'
make: *** [Makefile:852: test] Error 2
Command exited with non-zero status 2
	Command being timed: "/nfs/optimizer/tools/coin-or/coinbrew build Ipopt -j 4 --verbosity=4 --prefix=/home/russell/test-install/Ipopt/3.14.17/no-mumps --build-dir /home/russell/nfs-builds/Ipopt/3.14.17/no-mumps --disable-silent-rules ADD_CFLAGS=-pthread -fopenmp ADD_CXXFLAGS=-pthread -fopenmp -std=c++14 ADD_FCFLAGS=-fopenmp --with-lapack-lflags=-L/nfs/optimizer/intel/oneapi/mkl/2025.1/lib -L/nfs/optimizer/intel/oneapi/compiler/2025.1/lib  -lmkl_rt -liomp5 -Wl,-rpath -Wl,/nfs/optimizer/intel/oneapi/mkl/2025.1/lib:/nfs/optimizer/intel/oneapi/compiler/2025.1/lib --with-lapack-cflags=-I/nfs/optimizer/intel/oneapi/mkl/2025.1/include -m64 --without-mumps --without-hsl"
	User time (seconds): 66.16
	System time (seconds): 9.25
```

Again, removing `--without-mumps` provides a successful build:
```
Install completed. If executing any of the installed
binaries results in an error that shared libraries cannot
be found, you may need to
  - add 'export LD_LIBRARY_PATH=/home/russell/test-install/Ipopt/3.14.17/no-mumps/lib' to your ~/.bashrc (Linux)
  - add 'export DYLD_LIBRARY_PATH=/home/russell/test-install/Ipopt/3.14.17/no-mumps/lib' to ~/.bashrc (OS X)

	Command being timed: "/nfs/optimizer/tools/coin-or/coinbrew build Ipopt -j 4 --verbosity=4 --prefix=/home/russell/test-install/Ipopt/3.14.17/no-mumps --build-dir /home/russell/nfs-builds/Ipopt/3.14.17/no-mumps --disable-silent-rules ADD_CFLAGS=-pthread -fopenmp ADD_CXXFLAGS=-pthread -fopenmp -std=c++14 ADD_FCFLAGS=-fopenmp --with-lapack-lflags=-L/nfs/optimizer/intel/oneapi/mkl/2025.1/lib -L/nfs/optimizer/intel/oneapi/compiler/2025.1/lib  -lmkl_rt -liomp5 -Wl,-rpath -Wl,/nfs/optimizer/intel/oneapi/mkl/2025.1/lib:/nfs/optimizer/intel/oneapi/compiler/2025.1/lib --with-lapack-cflags=-I/nfs/optimizer/intel/oneapi/mkl/2025.1/include -m64 --without-hsl"
	User time (seconds): 101.84
	System time (seconds): 13.48
	Percent of CPU this job got: 284%

```
The failure is linking using libmkl_intel_thread because it is ignoring/not recognizing --with-lapack-lflags.

**Comment by svigerske** [comment]
The problem is actually not during the build, but when running the tests in Ipopt. I could not reproduce an issue when using `-lgomp`. With `-liomp5`, I could get an issue about finding `libiomp5.so`.
The `rpath` flags in `--with-lapack-lflags` are passed on when building the Ipopt library, but they are not used when building the tests (because they do not need Lapack themselves).

One option to bake the rpaths into the executables of the tests is to pass
```
LDFLAGS="-Wl,-rpath -Wl,/nfs/optimizer/intel/oneapi/mkl/2025.1/lib:/nfs/optimizer/intel/oneapi/compiler/2025.1/lib"
```

There is also `coinbrew` option `--tests none` to disable the tests.

**Comment by amontoison** [comment]
I close this issue because I think it is can of worms to isolate the symbols of `Pardiso-MKL` from the `BLAS` / `LAPACK` ones.
